test/test_c_cread.c

```diff
@@ -0,0 +1,183 @@
+/* This is a test for NCEPLIBS-bufr.
+ *
+ * This file tests misc C API stuff.
+ *
+ * Ed Hartnett 3/25/23 */
+
+#include <stdio.h>
+#include "bufrlib.h"
+#include "cread.h"
+
+#define FILENAME "test_c_cread.data"
+#define BAD_FILENAME "BAD_test_c_cread.data"
+#define IN_9 "testfiles/IN_9"
+#define TEST_ERR 1
+#define MSG_SIZE 14928 /* in bytes */
+
+void arallocc_(void);
+
+int main()
+{
+    int *msg, *msg2;
+    int old_val, i;
+    
+    printf("Testing cread functions.\n");
+
+    /* Allocate internal arrays. */
+    arallocc_();
+
+    /* Create a file. */
+    openwb(0, FILENAME);
+    if (!pb[0])
+        return TEST_ERR;
+    closfb(0);
+
+    /* Reopen the (empty) file for reading. */
+    openrb(0, FILENAME);
+    if (!pb[0])
+        return TEST_ERR;
+    closfb(0);
+
+    /* Reopen the (empty) file for appending. */
+    openab(0, FILENAME);
+    if (!pb[0])
+        return TEST_ERR;
+    closfb(0);
+
+    /* Allocate storage to read a BUFR message into. Without the +8
+     * these buffers will be overflowed by the cdrbufr(). See
+     * https://github.com/NOAA-EMC/NCEPLIBS-bufr/issues/450. */
+    if (!(msg = malloc(MSG_SIZE + 8)))
+        return TEST_ERR;
+    if (!(msg2 = malloc(MSG_SIZE + 8)))
+        return TEST_ERR;
+        
+    /* Open a real BUFR file, read a BUFR message, and close the file. */
```

Comment:
As I noted in the discussion of #450, I really don't think there's a need to do this.  I think the problem here is that you're confusing integers and bytes.  The msg and msg2 arrays are correctly declared as integer arrays, but the 3rd argument to crdbufr (i.e. mxwrd) is the number of **integer** elements of the array, not the number of bytes.  So instead of passing in MSG_SIZE as the 3rd argument to crdbufr, you should really be passing in MSGSIZE/4 (or MSGSIZE/sizeof(int), as below).

Reply:
If you do that (without allocating the 8 extra bytes) and it still fails, can you let me know what the return code from crdbufr is?  If it's -3 then I think what I'll do is tweak the logic to only check for nintrem+3>mxwrd if nbytrem%nb>0.  In other words, if the BUFR message size in bytes is an exact multiple of 4 (as in the case of 14928), then we only really need to check for nintrem+2>mxwrd as the overflow case, and maybe that's what's really going on here.

Let me know, but again in any case I don't think you should ever have to purposely allocate more space than you actually need to fit the largest size BUFR message you want to read.

Reply:
I tried your change of checking nbytrem%nb. This allows the function to proceed, but address sanitizer still tells me I am overflowing here:

```
    bufr[nintrem + 1] = wkint[0];
    bufr[nintrem + 2] = wkint[1];

```
Not sure why.

Reply:
As noted, I'll take a closer look at this in #450